4단계 - 로또(2등)

seogineer opened the pull request:
코드 리뷰 부탁드립니다.

seogineer committed:
refactor: containsKey 메소드로 isHitThree, isHitFour, isHitFive, isHitSix 대체

seogineer committed:
refactor: WinningNumber 클래스를 LottoNumber 클래스로 합침

seogineer committed:
docs: step4 기능 목록 작성

seogineer committed:
feat: 4단계 미션 문서에 enum Rank 소스에 따라 Prize 소스 수정

seogineer committed:
docs: Modify LottoNumber

seogineer committed:
docs: 지난주 당첨 번호로 입력했던 번호 입력 불가

seogineer committed:
docs: 보너스볼 번호 변수, 보너스볼 번호 포함 여부 메소드로 변경

seogineer committed:
docs: prize 변수 key값을 Integer 타입에서 Rank 타입으로 변경

- private final Map<Integer, Integer> prize = new HashMap<>();
- private final Map<Rank, Integer> prize = new HashMap<>();

Rank의 valueOf() 메소드와 사용하기 위해서 변경

seogineer committed:
refactor: prize 변수 key값을 Integer 타입에서 Rank 타입으로 변경

- private final Map<Integer, Integer> prize = new HashMap<>();
- private final Map<Rank, Integer> prize = new HashMap<>();

Rank의 valueOf() 메소드와 사용하기 위해서 변경

seogineer committed:
feat: 보너스볼 번호 변수, 보너스볼 번호 포함 여부 메소드 추가

seogineer committed:
feat: 보너스 볼 입력 및 결과 출력 추가

- Input View: 보너스 볼 입력 문구 추가
- Result View: 보너스 볼 당첨 통계 결과 출력 추가

seogineer committed:
docs: 유효성 검사 추가

- 지난주 당첨 번호 중에 중복된 번호가 있는지 검사
- 지난주 당첨 번호에 입력했던 번호 입력 불가

seogineer committed:
docs: 유효성 검사 추가

- 지난주 당첨 번호 중에 중복된 번호가 있는지 검사
- "로또 번호는 중복될 수 없습니다." 추가
- 지난주 당첨 번호에 입력했던 번호 입력 불가
- "입력하신 보너스 번호는 지난주 당첨 번호에 이미 있는 번호입니다." 추가

seogineer committed:
feat: 유효성 검사 추가

- 지난주 당첨 번호 중에 중복된 번호가 있는지 검사
- "로또 번호는 중복될 수 없습니다." 추가
- 지난주 당첨 번호에 입력했던 번호 입력 불가
- "입력하신 보너스 번호는 지난주 당첨 번호에 이미 있는 번호입니다." 추가

seogineer committed:
feat: 보너스 번호 입력 및 유효성 검사 추가

seogineer committed:
test: 테스트 코드 추가

geunhwanlee commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -9,15 +9,32 @@
 
     public LottoNumber(List<Integer> sixNumbers) {
         this.sixNumbers = sixNumbers;
+        this.bonusNumber = 0;
```
`보너스번호`에 의미없는 값으로 초기화 하고있는것으로 보이는데
`보너스번호`를 별개의 객체로 분리하는것은 어떨까요? 🤔

그리고 `로또`를 구성하는 여섯개의 각각의 `번호`와 `보너스번호`는 같은 특징을 가질것 같은데 어떻게 생각하시나요?(1~45 사이의 숫자만 가능함)

geunhwanlee commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,36 @@
+    public static Rank valueOf(int countOfMatch, boolean matchBonus) {
+        return Arrays.stream(values())
+                .filter(rank -> rank.countOfMatch == countOfMatch)
+                .filter(rank -> !rank.equals(SECOND) || matchBonus)
```
`countOfMatch`와 `matchBonus` 두개를 식별자로 `Rank`의 값들을 구성해보는건 어떨까요?🤔
`Rank`에 `matchBonus` 필드를 추가하면 될것 같은데 어떻게 생각하시나요?

geunhwanlee commented on src/main/java/lotto/Statistic.java:
```diff
@@ -12,89 +12,68 @@
     public void countPrize(List<LottoNumber> lottoNumbers) {
         for (LottoNumber lottoNumber : lottoNumbers) {
-            inputCountPrize(winningNumber.countHit(lottoNumber));
+            inputCountPrize(winningNumber.countHit(lottoNumber), winningNumber.containsBonus(lottoNumber));
```
`등수`를 만들기 위해

> 1. countHit의 결과를 받는다
> 2. containsBonus를 호출한 결과를 받는다
> 3. `1`, `2`번에서 받은 결과를 Rank.valueOf의 파라미터로 전달하여 `등수`를 받는다

의 프로세스를 수행하고있는데

이 역할을 `LottoNumber`에게 메세지를 보내 `등수`를 리턴받도록 하는건 어떤가요?🤔

geunhwanlee commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,36 @@
+                .orElse(MISS);
+    }
+
+    public int getCountOfMatch() {
+        return countOfMatch;
+    }
```
메서드를 사용하는곳이 없네요!
지워도 될듯 합니다.

geunhwanlee commented on src/main/java/lotto/Validate.java:
```diff
@@ -75,4 +80,25 @@ static void validateWinningNumberCount(String input) {
+        return checkSet.size();
+    }
+
+    static void validateBonusNumberDuplicate(int input, List<Integer> winningNumber) {
+        if (winningNumber.contains(input)) {
+            throw new IllegalArgumentException(ERROR_BONUS_NUMBER_DUPLICATED);
+        }
+    }
```
보너스볼 예외처리 👍🏻

geunhwanlee reviewed (changes requested):
4단계 미션 잘 해주셨어요!👏

몇가지 아쉬운 부분이 있어서 코멘트를 남겼어요.

기능을 구현하기 위한 역할들이 있는데
어떤 객체에게 역할을 위임해야할지
그렇게 하면 어떤 이점이 있는지, 또는 어떤 단점이 있는지를 고민해 보시면 좋을것 같아요!

한번 리팩토링 해보시고 재요청 주세요!🙏
질문사항 있으시면 언제든 환영입니다 🙌

그럼 화이팅하세요🔥

seogineer committed:
refactor: LottoNumber -> LottoNumbers로 클래스명 변경

seogineer committed:
refactor: 리뷰어님의 변경 요청 사항 반영

- 보너스 번호를 별개의 LottoNumber 클래스로 변경
- enum Rank에 matchBonus 필드 추가
- LottoNumbers 클래스에서 Rank를 받는 메세지 추가
- getCountOfMatch() 사용

geunhwanlee commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,42 @@
+
+    public static Rank valueOf(int countOfMatch, boolean matchBonus) {
+        return Arrays.stream(values())
+                .filter(rank -> rank.countOfMatch == countOfMatch)
+                .filter(rank -> !rank.equals(SECOND) || rank.matchBonus == matchBonus)
```
```suggestion
                .filter(rank -> rank.countOfMatch == countOfMatch && rank.matchBonus == matchBonus)
```
이렇게 하는건 어떤가요?🤔

geunhwanlee reviewed (approved):
리뷰내용 반영 잘해주셨어요!👍🏻

이번 단계는 여기서 머지하도록 할게요.

그럼 마지막 단계도 화이팅하세요!🔥
